### DISCUSSION TITLE
Define tiers of plugins for HLS

### DISCUSSION THREAD

**Discussion by michaelpj** [haskell/haskell-language-server#3123] [Ideas]
It would be helpful to identify certain HLS plugins as more important or better supported than others. For example:
- We advertise the support for newer GHC versions as "partial", meaning that not all plugins work for that version. But as we get more and more plugins, not all of which have highly active maintainers, "all plugins work" may become an unachievable goal. Indeed, the newly added `stan` plugin may never support GHCs other than 8.10! 
- Well-Typed is dong some work to update HLS for 9.4. We're asking them to continue that so that some plugins are supported... but which ones? It would be nice to have a clear set of "core" plugins that we really care about.

Here's a proposal. We define three tiers of plugin support:
1. Tier 1. These are "core" plugins, without which we don't think HLS is worth using. e.g. completions, hovers. This mostly corresponds to the `ghcide` plugins today, maybe a few more?
2. Tier 2. These are additional plugins which are well-maintained and quickly get updated to support new versions of GHC.
3. Tier 3. Everything else.

This interacts with GHC support as follows:
- We should not release a version of HLS supporting GHC X unless all Tier 1 plugins work. 
    - All Tier 1 plugins work = "basic support for GHC X"
- We should _try_ not to release a version of HLS supporting GHC X unless all Tier 2 plugins work, but we can if we need to.
    - All Tier 2 plugins work = "support for GHC X"
    - If a plugin fails to meet the bar for some period of time we should demote it to Tier 3.
- Tier 3 plugins are supported at a best-effort basis depending on their maintainers.
    - We don't say anything different about our support for GHC X based on which Tier 3 plugins work.

Tentative tier lists:
- Tier 1
    - ghcide plugins
    - `hls-call-hierarchy-plugin`
    - `hls-code-range-plugin`
    - `hls-pragmas-plugin`
    - `hls-explicit-imports-plugin`
- Tier 2
    - `hls-alternate-number-format-plugin`
    - `hls-chnage-type-signature-plugin`
    - `hls-class-plugin`
    - `hls-eval-plugin`
    - `hls-module-name-plugin`
    - `hls-hlint-plugin`
    - `hls-ormolu-plugin`
    - `hls-fourmolu-plugin`
    - `hls-stylish-haskell-plugin`
    - `hls-qualify-imported-names-plugin`
    - `hls-refine-imports-plugin`
    - `hls-rename-plugin`
    - `hls-gadt-plugin`
- Tier 3
    - `hls-brittany-plugin`    
    - `hls-haddock-comments-plugin`
    - `hls-stan-plugin`
    - `hls-retrie-plugin`
    - `hls-splice-plugin`
    - `hls-tactics-plugin`

My rubric here was basically: if it provides a relatively core LSP feature and it's fairly stable, it goes in Tier 1; if it's not currently up-to-date with the latest GHC we support, then clearly that's hard and it goes in Tier 3; everything else in Tier 2.

Thoughts?

**Comment by drsooch** [comment]
`hls-change-type-signature-plugin` and `alternate-number-format-plugin` can go in tier 2?

I would think `refine-imports` should be a Tier 1 feature, it seems like it would be the dual to `explicit-imports`, just my opinion

**Reply by michaelpj** [reply]
I think `refine-imports` is just not quite so established at the moment.

**Comment by pepeiborra** [comment]
I would bump `hls-eval-plugin` to Tier 1, under the premise that that's where it belongs even if its currently half-working or under-maintained. What do others think?

**Reply by michaelpj** [reply]
I think it's better for the tiers to reflect reality rather than aspiration. Otherwise we wouldn't be able to say we had more than basic support for 9.2, just because the eval plugin is not going anywhere. I think that's more misleading than the alternative, which is to say that the eval plugin is tier 2 but we'd like it to be tier 1. That suggests some concrete actions: can we find more maintainers, can we make it easier to upgrade?

Personally, I wouldn't be surprised if the eval plugin stays stuck in tier 2 since it's pretty complex and not *so* core to the HLS experience.

**Reply by michaelpj** [reply]
I guess I was also thinking that there is also some kind of difficulty modifier. We don't want things in tier 1 that are (in practice) difficult to update, even if they're very nice features.

**Reply by pepeiborra** [reply]
> to say that the eval plugin is tier 2 but we'd like it to be tier 1 ... suggests some concrete actions: can we find more maintainers, can we make it easier to upgrade?

Good point, agreed

**Comment by July541** [comment]
We added `hls-explicit-fixity-plugin` recently, I think it belongs to tier 2.

**Comment by georgefst** [comment]
I'm strongly in support of this, as I've seen multiple examples ([e.g. today on Reddit](https://www.reddit.com/r/haskell/comments/x4ot3e/comment/imxutol/?utm_source=share&utm_medium=web2x&context=3)) of people assuming that GHC 9.2 isn't supported, despite it being so almost-fully for months.

**Comment by michaelpj** [comment]
PR: https://github.com/haskell/haskell-language-server/pull/3189
